### PR TITLE
JAMES-2813 Task and additional information serialization contracts

### PR DESCRIPTION
**Pull request by chibenwa**
See this remark: https://github.com/linagora/james-project/pull/2932#discussion_r351782208

```
it should be a law like we do for Bean contract.
We should not need to write the same set of tests for each DTO type.
```

In this PR I contribute a common set of test. It will be applied as a sidecar refactoring along work on https://github.com/linagora/james-project/pull/2932 .

### CODE REVIEW

File: server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRestoreTaskSerializationTest.java [hunk 1]
```diff
@@ -18,84 +18,86 @@
...
-    void additionalInformationShouldBeSerializable() throws JsonProcessingException {
-        assertThatJson(JSON_TASK_ADDITIONAL_INFORMATION_SERIALIZER.serialize(DETAILS)).isEqualTo(SERIALIZED_ADDITIONAL_INFORMATION_TASK);
+    @Override
+    public TaskDTOModule taskDtoModule() {
+        return DeletedMessagesVaultRestoreTaskDTO.module(factory);
+    }
+
+    @Override
+    public String[] comparisonFields() {
+        return new String[0];
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRestoreTaskSerializationTest.java]: btw you overload it even when not needed ;)

File: server/task/task-json/src/test/java/org/apache/james/server/task/json/TaskAdditionalInformationSerializationContract.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import org.apache.james.server.task.json.dto.AdditionalInformationDTOModule;
+import org.apache.james.task.TaskExecutionDetails;
+import org.junit.jupiter.api.Test;
+
+public interface TaskAdditionalInformationSerializationContract {
+
+    String serializedAdditionalInformation();
+
+    TaskExecutionDetails.AdditionalInformation additionalInformation();
```
Comment (remk) [inline server/task/task-json/src/test/java/org/apache/james/server/task/json/TaskAdditionalInformationSerializationContract.java]: Maybe we could have a list o pair of serialized additional information and additional information.
And make the following test parameterized based on it ?

File: server/task/task-json/src/test/java/org/apache/james/server/task/json/TaskSerializationContract.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.james.server.task.json.dto.TaskDTOModule;
+import org.apache.james.task.Task;
+import org.junit.jupiter.api.Test;
+
+public interface TaskSerializationContract {
+    String serializedTask();
+
+    Task task();
```
Comment (remk) [inline server/task/task-json/src/test/java/org/apache/james/server/task/json/TaskSerializationContract.java]: idem ?

File: server/task/task-json/src/test/java/org/apache/james/server/task/json/TaskAdditionalInformationSerializationContract.java (hunk 2 above)
Comment (chibenwa) [reply server/task/task-json/src/test/java/org/apache/james/server/task/json/TaskAdditionalInformationSerializationContract.java]: Well, @ParameterizedTest requires static qualifier. Won't be easy to do IMO

**Comment by chibenwa** [comment]
> It's not bad but too specialized in mu opinion. I would rather build something like BeanVerifier: a tool you can call in a test easily.

Let's get a try at it!

**Comment by remk** [comment]
> > It's not bad but too specialized in mu opinion. I would rather build something like BeanVerifier: a tool you can call in a test easily.
> 
> Let's get a try at it!
> 
> ```
> JsonSerializationVerifier
>     .forBean(bean)
>     .associatedJson(json)
>     .serializationOperation(bean -> json)
>     .deserializationOperation(json -> bean)
>     .verify();
> ```

Or maybe a serialization verifier on a  `DTOModule` ?

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/3007

**Pull request closed without merging**